src/coprocessor/dag/expr/builtin_encryption.rs

```diff
@@ -91,13 +95,87 @@ impl ScalarFunc {
         };
         Ok(Some(Cow::Owned(sha2)))
     }
+
+    #[inline]
+    pub fn compress<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        if input.is_empty() {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        let mut e = ZlibEncoder::new(input.as_ref(), Compression::default());
+        let mut vec = Vec::with_capacity(1024);
+        match e.read_to_end(&mut vec) {
+            Ok(len) => {
+                let mut cap = 4 + len;
+                if vec[len - 1] == 32 {
+                    vec.push(b'.');
+                    cap += 1;
+                }
+
+                let mut wtr = Vec::with_capacity(cap);
+                wtr.write_u32::<LittleEndian>(input.len() as u32).unwrap();
+                wtr.extend_from_slice(&vec);
+                Ok(Some(Cow::Owned(wtr)))
+            }
+            _ => Ok(None),
+        }
+    }
+
+    #[inline]
+    pub fn uncompress<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        if input.is_empty() {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        if input.len() <= 4 {
+            ctx.warnings.append_warning(Error::zlib_data_corrupted());
+            return Ok(None);
+        }
+
+        let len = LittleEndian::read_u32(&input[0..4]) as usize;
+        let mut decoder = ZlibDecoder::new(&input[4..]);
+        let mut vec = Vec::with_capacity(len);
+        match decoder.read_to_end(&mut vec) {
+            Ok(decoded_len) if len >= decoded_len && decoded_len != 0 => Ok(Some(Cow::Owned(vec))),
```

Comment:
Can you explain the logic behind these comparisons?

Reply:
I some comments to explain these.

In the current implementation of TiDB, the first four bytes length is ignored , it try to uncompress , if success, return the string

In MySQL, if the first four bytes length is less than the actual 'uncompressed string' , it will return null and generate a warning `ZLIB: Not enough room in the output buffer (probably, length of uncompressed data was corrupted)`. Maybe it only read exact bytes of data, so caused corruped.

Reply:
Nice catch! Would you also like to file an issue in TiDB?